<a id="Python Basics"></a>
# Python Basics

## Entering expressions

In [14]:
2+2

4

**Commonly used math operators**
- Addition `+`
- Subtraction  `-`
- Multiplication `*`
- Division `/`
- Integer Division `//`
- Moduls/ Remainder `%`
- Exponent/ Power `**`

**Order of Precedence**

The order of operations (also called precedence) of Python math operators is similar to that of mathematics. The ** operator is evaluated first; the *, /, //, and % operators are evaluated next, from left to right; and the + and - operators are evaluated last (also from left to right)

In [15]:
2+3*6

20

In [16]:
(2 + 3) * 6

30

In [17]:
2**5

32

In [18]:
22/7

3.142857142857143

In [19]:
22//7

3

In [20]:
22 % 7

1

## Common data types

**Common Data Types**
- Integer - `int`
    - e.g. `0,50,-100`
- Floating-Point - `float`
    - e.g. `1.25, 3.14, -1.0, --0.5, 0.0`
- String - `str`
    - e.g. `'a', 'aa', "aaa", 'Hello!', '11 cats'`
- Boolean - `bool`
    - e.g. `True, False`

String can be enclosed in single or double quotes

## String concatenation & Replicaiton

The meaning of an operator (e.g. `+`) may change based on the data types of the values next to it


In [21]:
# Addition for int or float
1 + 2, 1.2 + 2.4

(3, 3.5999999999999996)

In [22]:
# Concatenation for string data type
'Bruce' + 'Wayne'

'BruceWayne'

 <p style="color:red;">If you try to use the + operator on a string and an integer value, Python will not know how to handle this, and it will display an error message</p>

In [23]:
'Bruce' + 42

TypeError: must be str, not int

The `*` operator when used on one string value and one integer value becomes string replication operator.

In [24]:
2*5

10

In [25]:
'Batman'*5

'BatmanBatmanBatmanBatmanBatman'

**Varaibles**
- It's like box in the computer’s memory where we can store values

**Variable Names** are case sensitive.<br>We  can name a variable anything as long as it obeys the following three rules:

- It can be only one word.
- It can use only letters, numbers, and the underscore `_` character.
- It can’t begin with a number.

e.g. of *invalid variable names*: 	
`current-balance, current balance 4account, total_$um, 'batman'`

## `Hello world` - First Python program

Let's write our first python program. The explanation for each line of code is given below as well.

In [38]:
# This program says hello and asks for my name.

print('Hello world!')
print('What is your name?')    # ask for their name
myName = input()
print('It is good to meet you, ' + myName)
print('The length of your name is:')
print(len(myName))
print('What is your age?')    # ask for their age
myAge = input()
print (type(myAge))
print('You will be ' + str(int(myAge) + 1) + ' in a year.')

Hello world!
What is your name?
Super Man
It is good to meet you, Super Man
The length of your name is:
9
What is your age?
40
<class 'str'>
You will be 41 in a year.


### Program Breakdown

#### Comments

The following line is called a comment. Python ignores comments, and you can use them to write notes or remind yourself what the code is trying to do. Any text for the rest of the line following a hash mark `#` is part of a comment.

In [27]:
# This program says hello and asks for my name.

To comment across multiple lies, we can use triple quotes (could be single `'''` or double `"""`)

In [28]:
'''This line is part of multine comment

This line is also part of multiline comment

Again, this line as well is a multiline comment'''

print ('Hello')

Hello


#### `print()` function

The `print()` function displays the `string` value inside the parentheses on the screen. Notice that the quotes are not printed to the screen. They just mark where the string begins and ends; they are not part of the string value.

In [39]:
print('Hello world!')
print('What is your name?')    # ask for their name

Hello world!
What is your name?


#### `input()` function

The `input()` function waits for the user to type some text on the keyboard and press ENTER key.
The user input will then be stored in `myName` variable

In [40]:
myName = input()

Super Man


The expression `'It is good to meet you, ' + myName` is evaluated and is `print`ed on the screen.

In [41]:
print('It is good to meet you, ' + myName)

It is good to meet you, Super Man


#### `len()` function

The `len()` function returns the number of characters in the given string.

In [42]:
print('The length of your name is:')
print(len(myName))

The length of your name is:
9


#### type() function

The `type()` function returns the data type of the value given

In [44]:
print('What is your age?')    # ask for their age
myAge = input()
print (type(myAge))           # the input functions usually returns a string data type

What is your age?
40
<class 'str'>


#### `str(), int(), float()` functions

These functions are used to change the data types of the values. <br>
As the `input()` function returns `str` data, `myAge` variable is of `str` data type. <br>
To add a numerical value to it, the value should be converted to `int`.<br> 
After performing the addition, the `int` value is converted back to `str` again so that it can be concatenated with other strings.

In [45]:
print('You will be ' + str(int(myAge) + 1) + ' in a year.')

You will be 41 in a year.


# Task  1 

Now, let's take a break from learning and start testing what we have learnt. 

- Open <code><font style="color:#FF8b33">1-Python-Notes-Exercise</font></code>  > `3. Create shapes` and follow the instructions. 
- If you are new to python `turtle`, read upon `import turtle` and `Move the turtle` written on top of the same page. 

Upon completion, return back to this section.

To view and download the *tasks* in `HTML` : <a href="https://prasanth-ntu.github.io/html/learn-python/1-Python-Notes-Exercise.html" target="_blank">Click here</a> and in `GITHUB` : <a href="https://prasanth-ntu.github.io/html/learn-python/1-Python-Notes-Exercise.ipynb" target="_blank">Click here</a>